core/src/main/scala/kafka/coordinator/transaction/TransactionCoordinator.scala

```diff
@@ -540,24 +572,29 @@ class TransactionCoordinator(txnConfig: TransactionConfig,
                   txnMetadata.lastProducerEpoch = RecordBatch.NO_PRODUCER_EPOCH
                 }
 
-                Right(coordinatorEpoch, txnMetadata.prepareAbortOrCommit(nextState, time.milliseconds()))
+                nextProducerIdOrErrors match {
+                  case Left(error) =>
+                    Left(error)
+                  case Right(nextProducerId) =>
+                    Right(coordinatorEpoch, txnMetadata.prepareAbortOrCommit(nextState, clientTransactionVersion, nextProducerId, time.milliseconds()))
+                }
               case CompleteCommit =>
```

Comment:
Could refactor this with isInvalidTxnTransition

Reply:
Originally I put the retry check above the epoch check since I thought it would be easier to do that first and not touch the check for the same epoch. The tricky thing is specifically that if we hit this code we have the same epoch which is wrong. (In other words, we should have the same OR a different epoch depending on the state -- but I can factor that all in the epoch check, it just means we will check the state in two places.)